### DISCUSSION TITLE
Constrained relaxation and dynamics

### DISCUSSION THREAD

**Discussion by andrea-angeletti** [CederGroupHub/chgnet#103] [Q&A]
How can I perform structure optimisation and molecular dynamics by constraining the cell shape? In VASP it would be equivalent to ISIF=8 or what you can achieve with ICONST, which allows you to relax the lattice parameters without changing the angle between primitive vectors.

**Comment by bowen-bd** [comment]
This is currently not implemented.
And I looked into [ase.filters](https://gitlab.com/ase/ase/-/blob/master/ase/filters.py) and [ase.constraints](https://gitlab.com/ase/ase/-/blob/master/ase/constraints.py), seems there is no support for this neither.

The easist way to implement this is to force the stress tensor to be diagonal before [passing to a structural optimization.](https://github.com/CederGroupHub/chgnet/blob/db40e05eaaa9b292a31877e713ba328d3365ebcf/chgnet/model/dynamics.py#L144)

Can you please test this and make a pull request if it works well?

**Reply by janosh** [reply]
There is [`FixSymmetry`](https://wiki.fysik.dtu.dk/ase/ase/constraints.html#the-fixsymmetry-class). Never used it myself but maybe that does what you want.

**Reply by janosh** [reply]
Oh and also [`FixInternals`](https://wiki.fysik.dtu.dk/ase/ase/constraints.html#the-fixinternals-class).

> Allows fixing bonds, angles, dihedrals as well as linear combinations of bonds (bondcombos).

**Reply by bowen-bd** [reply]
@janosh I suppose `FixSymmetry` also fix the atoms to be on high-symmetry points

**Reply by janosh** [reply]
Based on doc str, you can control angle fixing and atom fixing separately.
